core/lib/snapshots_applier/src/lib.rs

```diff
@@ -0,0 +1,282 @@
+        SnapshotStorageLogsChunk, SnapshotStorageLogsStorageKey,
+    },
+    zkevm_test_harness::k256::pkcs8::der::EncodeValue,
+    MiniblockNumber, H256,
+};
+use zksync_utils::bytecode::hash_bytecode;
+use zksync_web3_decl::{
+    jsonrpsee::core::ClientError as RpcError,
+    namespaces::{EthNamespaceClient, ZksNamespaceClient},
+};
+
+use crate::SnapshotApplierError::*;
+
+mod tests;
+
+#[derive(thiserror::Error, Debug)]
+pub enum SnapshotApplierError {
+    #[error("canceled")]
+    Canceled(String),
+    #[error(transparent)]
+    Fatal(#[from] anyhow::Error),
+    #[error("retryable")]
+    Retryable(String),
+}
+pub struct SnapshotsApplier<'a> {
+    connection_pool: &'a ConnectionPool,
+    blob_store: Box<dyn ObjectStore>,
+    applied_snapshot_status: SnapshotRecoveryStatus,
+}
+
+impl From<ObjectStoreError> for SnapshotApplierError {
+    fn from(error: ObjectStoreError) -> Self {
+        match error {
+            ObjectStoreError::KeyNotFound(err) => {
+                Fatal(Error::msg(format!("Key not found in object store: {err}")))
+            }
+            ObjectStoreError::Serialization(err) => {
+                Fatal(Error::msg(format!("Unable to deserialize object: {err}")))
+            }
+            ObjectStoreError::Other(_) => {
+                Retryable("An error occurred while accessing object store".to_string())
+            }
+        }
+    }
+}
+
+impl From<SqlxError> for SnapshotApplierError {
+    fn from(err: SqlxError) -> Self {
+        match err {
+            SqlxError::Database(_) => Fatal(err.into()),
+            SqlxError::RowNotFound => Fatal(err.into()),
+            SqlxError::ColumnNotFound(_) => Fatal(err.into()),
+            SqlxError::Configuration(_) => Fatal(err.into()),
+            SqlxError::TypeNotFound { type_name: _ } => Fatal(err.into()),
+            _ => Retryable(format!("An error occured when accessing DB: {err}")),
+        }
+    }
+}
+
+#[async_trait]
+pub trait SnapshotsApplierMainNodeClient: fmt::Debug + Send + Sync {
+    async fn miniblock_hash(&self, number: MiniblockNumber) -> Result<Option<H256>, RpcError>;
+
+    async fn fetch_newest_snapshot(&self) -> Result<Option<SnapshotHeader>, RpcError>;
+}
+impl<'a> SnapshotsApplier<'a> {
+    pub async fn load_snapshot(
+        connection_pool: &'a ConnectionPool,
+        main_node_client: Box<dyn SnapshotsApplierMainNodeClient>,
+        blob_store: Box<dyn ObjectStore>,
+    ) -> anyhow::Result<(), SnapshotApplierError> {
+        let mut storage = connection_pool
+            .access_storage_tagged("snapshots_applier")
+            .await?;
+        let mut storage = storage.start_transaction().await?;
+
+        let mut applied_snapshot_status = storage
+            .snapshot_recovery_dal()
+            .get_applied_snapshot_status()
+            .await
+            .map_err(|_| Retryable("Unable to get applied snapshots status".to_string()))?;
```

Comment:
You have `From<SqlxError> for SnapshotApplierError` conversion defined above; why is it not used here and in similar situations? If you want to add additional context for errors, one way to do this would be to define an extension trait similar to `anyhow::Context` that would convert `Result<T, E>` where `E: Into< SnapshotApplierError>` + context into `Result<T, SnapshotApplierError>`. Something like this:

```rust
trait SnapshotApplierContext<T> {
    fn wrap(self, context: &'static str) -> Result<T, SnapshotApplierError>;
    // Can define a lazily evaluated context (e.g., `impl FnOnce() -> String`) as another method
}

impl<T, E: Into<SnapshotApplierError>> SnapshotApplierContext<T> for Result<T, E> {
    fn wrap(self, context: &'static str) -> Result<T, SnapshotApplierError> {
        match self {
            Ok(value) => Ok(value),
            Err(err) => Err(err.into().context(context)),
        }
    }
}

impl SnapshotApplierError {
    fn context(self, context: &'static str) -> Self {
        // To be able to add context, all error variants should be `anyhow::Error`,
        // or it can be refactored to a struct with `anyhow::Error` and severity fields
    }
}

// usage:
let mut applied_snapshot_status = storage
    .snapshot_recovery_dal()
    .get_applied_snapshot_status()
    .await
    .wrap("Unable to get applied snapshots status")?;
```

Reply:
I've actually just forgot to refactor those places, thanks for providing the snippet, I plan to use it when adding resiliency